src/Engine/Objects/SpriteObject.cpp

```diff
@@ -351,235 +353,146 @@ void SpriteObject::UpdateObject_fn0_BLV(unsigned int uLayingItemID) {
         SpriteObject::OnInteraction(uLayingItemID);
         return;
     }
-    pSpriteObject->uSectorID = uSectorID;
-
-    int v15;               // ebx@46
-    int v17;                      // eax@50
-    int16_t v22;                  // ax@57
-    int v23;                      // edi@62
-    Particle_sw Dst;              // [sp+Ch] [bp-84h]@18
-
-    int v39;                      // [sp+80h] [bp-10h]@33
-    int v40;                      // [sp+84h] [bp-Ch]@28
 
+    pSpriteObject->uSectorID = uSectorID;
 
-    if (pObject->uFlags & OBJECT_DESC_NO_GRAVITY) {
-        goto LABEL_25;
+    if (!(pObject->uFlags & OBJECT_DESC_NO_GRAVITY)) {
+        if (floor_lvl <= pSpriteObject->vPosition.z - 3) {
+            pSpriteObject->vVelocity.z -= (short)pEventTimer->uTimeElapsed * GetGravityStrength();
+        }
     }
 
     // flying objects / projectiles
     if (floor_lvl <= pSpriteObject->vPosition.z - 3) {
-        pSpriteObject->vVelocity.z -= (short)pEventTimer->uTimeElapsed * GetGravityStrength();
 LABEL_25:
```

Comment:
I believe we have changed the logic here.

The logic before was:
1. If no gravity => go unconditionally into collisions.
2. If gravity => if we're above the floor, then apply gravity & go into collisions, else do nothing.

The logic now:
1. If no gravity => if we're above the floor, then go into collisions.
2. If gravity => if we're above the floor, then apply gravity & go into collisions, else do nothing.

I suggest:
1. Don't change the logic for now.
2. I believe collisions should be applied unconditionally to everything. Let's add a TODO here and fix this in a separate PR.

Reply:
Yeah gotos are pain in the ass